src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs

```diff
@@ -140,24 +164,51 @@ protected override void ProcessRecord()
                 if (_jschema != null)
                 {
-                    var errorMessages = _jschema.Validate(parsedJson);
-                    if (errorMessages != null && errorMessages.Count != 0)
+                    try
                     {
-                        result = false;
-
-                        Exception exception = new(TestJsonCmdletStrings.InvalidJsonAgainstSchema);
-
-                        foreach (var message in errorMessages)
+                        var validationResults = _jschema.Validate(parsedJson, new ValidationOptions { OutputFormat = OutputFormat.Basic });
+                        result = validationResults.IsValid;
+                        if (!result)
                         {
-                            ErrorRecord errorRecord = new(exception, "InvalidJsonAgainstSchema", ErrorCategory.InvalidData, null);
-                            errorRecord.ErrorDetails = new ErrorDetails(message.ToString());
-                            WriteError(errorRecord);
+                            Exception exception = new(TestJsonCmdletStrings.InvalidJsonAgainstSchema);
+
+                            if (validationResults.Message != null)
+                            {
+                                ErrorRecord errorRecord = new(exception, "InvalidJsonAgainstSchema",
+                                    ErrorCategory.InvalidData, null);
+                                var message = $"{validationResults.Message} at {validationResults.InstanceLocation}";
+                                errorRecord.ErrorDetails = new ErrorDetails(message);
+                                WriteError(errorRecord);
+                            }
+
+                            if (validationResults.HasNestedResults)
+                            {
+                                foreach (var nestedResult in validationResults.NestedResults)
+                                {
+                                    if (nestedResult.Message == null)
+                                    {
+                                        continue;
+                                    }
+
+                                    ErrorRecord errorRecord = new(exception, "InvalidJsonAgainstSchema", ErrorCategory.InvalidData, null);
+                                    var message = $"{nestedResult.Message} at {nestedResult.InstanceLocation}";
+                                    errorRecord.ErrorDetails = new ErrorDetails(message);
+                                    WriteError(errorRecord);
+                                }
+                            }
                         }
                     }
+                    catch (JsonException jsonExc)
+                    {
+                        result = false;
+
+                        Exception exception = new(TestJsonCmdletStrings.InvalidJsonSchema, jsonExc);
+                        WriteError(new ErrorRecord(exception, "InvalidJsonSchema", ErrorCategory.InvalidData, Json));
+                    }
```

Comment:
I wonder if Validate() throw on schema but not JsonSchema.FromFile() or JsonSchema.FromText().

Reply:
It would, but isolating the `.Validate()` is the only way to detect that it was a _referenced_ schema that failed to parse, so we still need this here.

Reply:
> it was a _referenced_ schema that failed to parse

Why isn't it parsed/validated at JsonSchema.FromText/FromFile time? I'd expect validate API never throw but return a result.
It may make sense to postpone this operation, but your API becomes somewhat difficult to understand, at least it violates expectations. In this context, it would make more sense if JsonSchemaException throws instead of JsonException.

Reply:
There's some confusion.

`.FromText()`/`.FromFile()` load the initial schema by parsing & deserializing a single JSON text.  A failure here would raise a `JsonException`.  This is happening in `BeginProcessing()`.

If that schema has references (`$ref` and `$dynamicRef`) that lead to external files/locations, the references are resolved at validation time, which means that they must be deserialized at validation time, but that also yields a `JsonException`.  To distinguish between the instance (the JSON data being validated) being bad JSON and trying to deserialize a bad schema, we have to isolate the `.Validate()` method.  If `.Validate()` throws a `JsonException`, we know it must be from an attempt to deserialize a referenced schema.

Reply:
I think we should add such comment in code.